Readme.md
Lua module for CRC-32 calculation implemented in C. 

Tested with Lua 5.1, 5.2 and 5.3.

Also included is a pure Lua implementation of the same API, using
https://github.com/davidm/lua-digest-crc32lua.


# Module API
`local crc32 = require 'crc32'`

* **crc32.crc32** = function (crc_in, data)
	* **crc_in** is input CRC. It can be either a Lua number, or a string.
	If string is used, it should be 4 bytes in big endian order. 
	If string shorter than 4 bytes, then CRC is left padded with zeros.
	* **data** is the input data to apply to CRC, as a Lua string.
	* **returns** updated CRC. If crc_in is number, then returns number. 
	If crc_in is string, then return as string, 4 bytes in big endian order.
* **crc32.newcrc32** = function ()
	* **returns** a stateful CRC-32 calculator object (userdata). 
	It is useful for streaming CRC calculation, as it keep the current
	CRC as a native uint32_t between calls, reducing conversion overhead. 
* **crc32.version** = string

## crc32.newcrc32 API
`local c = crc32.newcrc32()`

* **c.update** = function (self, data)
	* **data** is the input data to apply to CRC, as a Lua string.
	* **returns** self
* **c.reset** = function (self) 
	* Resets the current CRC to 0. 
	* **returns** self
* **c.tonumber** = function (self)
	* **returns** current CRC as Lua number
* **c.tostring** = function (self)
	* **returns** current CRC as string, 4 bytes in big endian order.
* **c.tohex** = function (self)
	* **returns** current CRC as 8 character hex encoded string.


debian/changelog
lua-crc32 (1.1-1) unstable; urgency=medium

  * Initial packaging

 -- Egil Hjelmeland <privat@egil-hjelmeland.no>  Mon, 16 Apr 2018 14:57:46 +0100


test_crc32.lua
#!/usr/bin/env lua
local _G,arg,assert,getmetatable,package,require,string,type
    = _G,arg,assert,getmetatable,package,require,string,type

local C32 = require'crc32'

if arg[-1] ~= '-lcrc32' then  -- work around debian dh_lua test calling with '-lcrc32', which sets _G.crc32
	assert(_G.crc32==nil, 'do not pollute global env')
end
assert(type(C32.version)=='string')
assert(type(C32.crc32)=='function')

do
	local crc32=C32.crc32

	assert(crc32(0, 'egil') == 0x901FF815)
	assert(crc32(0, 'eg') == 0x8A8579E6)
	assert(crc32(0x8A8579E6, 'il') == 0x901FF815)

	-- test with null byte
	assert(crc32(0, string.char (1,0) ) == 0x58C223BE)


	-- test with CRC as string
	assert ( crc32('', 'egil') == string.char (0x90, 0x1F, 0xF8, 0x15))
	local c1 = crc32('', 'eg')
	assert( c1 == string.char (0x8A, 0x85, 0x79, 0xE6))
	assert ( crc32(c1, 'il') == string.char (0x90, 0x1F, 0xF8, 0x15))
end

do -- test .newcrc32()

	local crccalc = C32.newcrc32()
	--assert(type(crccalc) == 'userdata')
	local crccalc_mt = getmetatable(crccalc)
	assert(crccalc_mt.reset)


	assert(crccalc:tonumber() == 0)
	assert(crccalc:tostring() == string.char (0, 0, 0, 0))
	assert(crccalc:tohex() == '00000000')
	
	assert(crccalc:update('egil' , 'surplussparameter') == crccalc )
	assert(crccalc:tonumber() == 0x901FF815)
	assert(crccalc:tostring() == string.char (0x90, 0x1F, 0xF8, 0x15))
	assert(crccalc:tohex() == '901ff815')

	assert(crccalc:reset('surplussparameter') == crccalc )
	assert(crccalc:tonumber() == 0)
	assert(crccalc:tohex() == '00000000')

	crccalc:update'eg':update'il'
	assert(crccalc:tohex() == '901ff815')
end


--[[ -- test performance
local S = require'socket'
local now = S.gettime()

local d ={}
for i = 1, 1000 do 
	d[i] = ('%d'):format(i+1000)
end

--------------------------
local now = S.gettime()
crc = 0
for i = 1, 1000 do 
	crc = crc32(crc, d[i])
end
print( (S.gettime()- now))
print(('%8.8x'):format(crc))
print()


--------------------------
local now = S.gettime()
crc = ''
for i = 1, 1000 do 
	crc = crc32(crc, d[i])
end
print( (S.gettime()- now))
print(crc:gsub('.', function(b) return ('%2.2x'):format(b:byte()) end) )
print()

--------------------------

local now = S.gettime()
crc =  C32.newcrc32()
for i = 1, 1000 do 
	crc:update( d[i])
end
print( (S.gettime()- now))
print(crc:tohex() )

--]]



crc32.c
/*-
 *  COPYRIGHT (C) 1986 Gary S. Brown.  You may use this program, or
 *  code or tables extracted from it, as desired without restriction.
 *
 *  First, the polynomial itself and its table of feedback terms.  The
 *  polynomial is
 *  X^32+X^26+X^23+X^22+X^16+X^12+X^11+X^10+X^8+X^7+X^5+X^4+X^2+X^1+X^0
 *
 *  Note that we take it "backwards" and put the highest-order term in
 *  the lowest-order bit.  The X^32 term is "implied"; the LSB is the
 *  X^31 term, etc.  The X^0 term (usually shown as "+1") results in
 *  the MSB being 1
 *
 *  Note that the usual hardware shift register implementation, which
 *  is what we're using (we're merely optimizing it by doing eight-bit
 *  chunks at a time) shifts bits into the lowest-order term.  In our
 *  implementation, that means shifting towards the right.  Why do we
 *  do it this way?  Because the calculated CRC must be transmitted in
 *  order from highest-order term to lowest-order term.  UARTs transmit
 *  characters in order from LSB to MSB.  By storing the CRC this way
 *  we hand it to the UART in the order low-byte to high-byte; the UART
 *  sends each low-bit to hight-bit; and the result is transmission bit
 *  by bit from highest- to lowest-order term without requiring any bit
 *  shuffling on our part.  Reception works similarly
 *
 *  The feedback terms table consists of 256, 32-bit entries.  Notes
 *
 *      The table can be generated at runtime if desired; code to do so
 *      is shown later.  It might not be obvious, but the feedback
 *      terms simply represent the results of eight shift/xor opera
 *      tions for all combinations of data and CRC register values
 *
 *      The values must be right-shifted by eight bits by the "updcrc
 *      logic; the shift must be unsigned (bring in zeroes).  On some
 *      hardware you could probably optimize the shift in assembler by
 *      using byte-swap instructions
 *      polynomial $edb88320
 *
 *
 * CRC32 code derived from work by Gary S. Brown.
 */

#include <stddef.h>
#include <stdint.h>

static uint32_t crc32_tab[] = {
	0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419, 0x706af48f,
	0xe963a535, 0x9e6495a3,	0x0edb8832, 0x79dcb8a4, 0xe0d5e91e, 0x97d2d988,
	0x09b64c2b, 0x7eb17cbd, 0xe7b82d07, 0x90bf1d91, 0x1db71064, 0x6ab020f2,
	0xf3b97148, 0x84be41de,	0x1adad47d, 0x6ddde4eb, 0xf4d4b551, 0x83d385c7,
	0x136c9856, 0x646ba8c0, 0xfd62f97a, 0x8a65c9ec,	0x14015c4f, 0x63066cd9,
	0xfa0f3d63, 0x8d080df5,	0x3b6e20c8, 0x4c69105e, 0xd56041e4, 0xa2677172,
	0x3c03e4d1, 0x4b04d447, 0xd20d85fd, 0xa50ab56b,	0x35b5a8fa, 0x42b2986c,
	0xdbbbc9d6, 0xacbcf940,	0x32d86ce3, 0x45df5c75, 0xdcd60dcf, 0xabd13d59,
	0x26d930ac, 0x51de003a, 0xc8d75180, 0xbfd06116, 0x21b4f4b5, 0x56b3c423,
	0xcfba9599, 0xb8bda50f, 0x2802b89e, 0x5f058808, 0xc60cd9b2, 0xb10be924,
	0x2f6f7c87, 0x58684c11, 0xc1611dab, 0xb6662d3d,	0x76dc4190, 0x01db7106,
	0x98d220bc, 0xefd5102a, 0x71b18589, 0x06b6b51f, 0x9fbfe4a5, 0xe8b8d433,
	0x7807c9a2, 0x0f00f934, 0x9609a88e, 0xe10e9818, 0x7f6a0dbb, 0x086d3d2d,
	0x91646c97, 0xe6635c01, 0x6b6b51f4, 0x1c6c6162, 0x856530d8, 0xf262004e,
	0x6c0695ed, 0x1b01a57b, 0x8208f4c1, 0xf50fc457, 0x65b0d9c6, 0x12b7e950,
	0x8bbeb8ea, 0xfcb9887c, 0x62dd1ddf, 0x15da2d49, 0x8cd37cf3, 0xfbd44c65,
	0x4db26158, 0x3ab551ce, 0xa3bc0074, 0xd4bb30e2, 0x4adfa541, 0x3dd895d7,
	0xa4d1c46d, 0xd3d6f4fb, 0x4369e96a, 0x346ed9fc, 0xad678846, 0xda60b8d0,
	0x44042d73, 0x33031de5, 0xaa0a4c5f, 0xdd0d7cc9, 0x5005713c, 0x270241aa,
	0xbe0b1010, 0xc90c2086, 0x5768b525, 0x206f85b3, 0xb966d409, 0xce61e49f,
	0x5edef90e, 0x29d9c998, 0xb0d09822, 0xc7d7a8b4, 0x59b33d17, 0x2eb40d81,
	0xb7bd5c3b, 0xc0ba6cad, 0xedb88320, 0x9abfb3b6, 0x03b6e20c, 0x74b1d29a,
	0xead54739, 0x9dd277af, 0x04db2615, 0x73dc1683, 0xe3630b12, 0x94643b84,
	0x0d6d6a3e, 0x7a6a5aa8, 0xe40ecf0b, 0x9309ff9d, 0x0a00ae27, 0x7d079eb1,
	0xf00f9344, 0x8708a3d2, 0x1e01f268, 0x6906c2fe, 0xf762575d, 0x806567cb,
	0x196c3671, 0x6e6b06e7, 0xfed41b76, 0x89d32be0, 0x10da7a5a, 0x67dd4acc,
	0xf9b9df6f, 0x8ebeeff9, 0x17b7be43, 0x60b08ed5, 0xd6d6a3e8, 0xa1d1937e,
	0x38d8c2c4, 0x4fdff252, 0xd1bb67f1, 0xa6bc5767, 0x3fb506dd, 0x48b2364b,
	0xd80d2bda, 0xaf0a1b4c, 0x36034af6, 0x41047a60, 0xdf60efc3, 0xa867df55,
	0x316e8eef, 0x4669be79, 0xcb61b38c, 0xbc66831a, 0x256fd2a0, 0x5268e236,
	0xcc0c7795, 0xbb0b4703, 0x220216b9, 0x5505262f, 0xc5ba3bbe, 0xb2bd0b28,
	0x2bb45a92, 0x5cb36a04, 0xc2d7ffa7, 0xb5d0cf31, 0x2cd99e8b, 0x5bdeae1d,
	0x9b64c2b0, 0xec63f226, 0x756aa39c, 0x026d930a, 0x9c0906a9, 0xeb0e363f,
	0x72076785, 0x05005713, 0x95bf4a82, 0xe2b87a14, 0x7bb12bae, 0x0cb61b38,
	0x92d28e9b, 0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21, 0x86d3d2d4, 0xf1d4e242,
	0x68ddb3f8, 0x1fda836e, 0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777,
	0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c, 0x8f659eff, 0xf862ae69,
	0x616bffd3, 0x166ccf45, 0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2,
	0xa7672661, 0xd06016f7, 0x4969474d, 0x3e6e77db, 0xaed16a4a, 0xd9d65adc,
	0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
	0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605, 0xcdd70693,
	0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
	0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
};

uint32_t
crc32(uint32_t crc, const void *buf, size_t size)
{
	const uint8_t *p;

	p = buf;
	crc = crc ^ ~0U;

	while (size--)
		crc = crc32_tab[(crc ^ *p++) & 0xFF] ^ (crc >> 8);

	return crc ^ ~0U;
}


crc32.lua
local  crc32lua = require 'crc32lua'
local bit = require'bit32'
local band = bit.band
local bor = bit.bor
local rshift = bit.rshift
local lshift = bit.lshift

local crc32_string =  crc32lua.crc32_string

local crc_mt = {
	update = function (self, data)
		self[1] =  crc32_string(data, self[1]  )
		return self
	end, 
	reset = function (self)
		self[1] = 0
		return self
	end, 
	tonumber = function (self)
		return self[1] 
	end, 
	tostring = function (self)
		local crc_out = self[1] 
		local crcout_bytes = {}
		for i =  4, 1,-1  do
			crcout_bytes[i] = string.char(band( crc_out , 0xff))
			crc_out = rshift(crc_out, 8)
		end
		return table.concat(crcout_bytes)
	end, 
	tohex = function (self)
		return ('%8.8x'):format(self[1] )
	end, 
}
crc_mt.__index = crc_mt


return {
	crc32 = function (crc, data)
	
		if type(crc) == 'string' then
			crc_len=#crc
			if crc_len>4 then crc_len = 4 end
			local crc_in = 0;
			for i =  1, crc_len do
				crc_in = lshift(crc_in, 8)
				crc_in = bor(crc_in, crc:byte(i))
			end
			
			local crc_out = crc32_string(data, crc_in )
			local crcout_bytes = {}
			for i =  4, 1,-1  do
				crcout_bytes[i] = string.char(band( crc_out , 0xff))
				crc_out = rshift(crc_out, 8)
			end
			return table.concat(crcout_bytes)
		end
		return crc32_string(data, crc )
	end,
	
	newcrc32 = function ()
		return setmetatable({0}, crc_mt)
	end,
	version='1.0',
}


wrap.c
#include <stddef.h>
#include <stdint.h>

#include "lua.h"
#include "lauxlib.h"

#define VERSION "1.00"

#if LUA_VERSION_NUM < 502
/* -----copied from bit32.c: ----- */

/* ----- adapted from lua-5.2.0 luaconf.h: ----- */

/*
@@ LUA_UNSIGNED is the integral type used by lua_pushunsigned/lua_tounsigned.
** It must have at least 32 bits.
*/
#define LUA_UNSIGNED    unsigned LUAI_INT32

#if defined(LUA_NUMBER_DOUBLE) && !defined(LUA_ANSI)    /* { */

/* On a Microsoft compiler on a Pentium, use assembler to avoid clashes
   with a DirectX idiosyncrasy */
#if defined(LUA_WIN) && defined(_MSC_VER) && defined(_M_IX86)	/* { */

#define MS_ASMTRICK

#else                           /* }{ */
/* the next definition uses a trick that should work on any machine
   using IEEE754 with a 32-bit integer type */

#define LUA_IEEE754TRICK

/*
@@ LUA_IEEEENDIAN is the endianness of doubles in your machine
** (0 for little endian, 1 for big endian); if not defined, Lua will
** check it dynamically.
*/
/* check for known architectures */
#if defined(__i386__) || defined(__i386) || defined(__X86__) || \
    defined (__x86_64)
#define LUA_IEEEENDIAN  0
#elif defined(__POWERPC__) || defined(__ppc__)
#define LUA_IEEEENDIAN  1
#endif

#endif                          /* } */

#endif                  /* } */

/* ----- from lua-5.2.0 lua.h: ----- */

/* unsigned integer type */
typedef LUA_UNSIGNED lua_Unsigned;

/* ----- adapted from lua-5.2.0 llimits.h: ----- */

/* lua_number2unsigned is a macro to convert a lua_Number to a lua_Unsigned.
** lua_unsigned2number is a macro to convert a lua_Unsigned to a lua_Number.
*/

#if defined(MS_ASMTRICK)        /* { */
/* trick with Microsoft assembler for X86 */

#define lua_number2unsigned(i,n)  \
  {__int64 l; __asm {__asm fld n   __asm fistp l} i = (unsigned int)l;}

#elif defined(LUA_IEEE754TRICK)         /* }{ */
/* the next trick should work on any machine using IEEE754 with
   a 32-bit integer type */

union luai_Cast2 { double l_d; LUAI_INT32 l_p[2]; };

#if !defined(LUA_IEEEENDIAN)    /* { */
#define LUAI_EXTRAIEEE  \
  static const union luai_Cast2 ieeeendian = {-(33.0 + 6755399441055744.0)};
#define LUA_IEEEENDIAN          (ieeeendian.l_p[1] == 33)
#else
#define LUAI_EXTRAIEEE          /* empty */
#endif                          /* } */

#define lua_number2int32(i,n,t) \
  { LUAI_EXTRAIEEE \
    volatile union luai_Cast2 u; u.l_d = (n) + 6755399441055744.0; \
    (i) = (t)u.l_p[LUA_IEEEENDIAN]; }

#define lua_number2unsigned(i,n)        lua_number2int32(i, n, lua_Unsigned)

#endif                          /* } */

#endif
#if LUA_VERSION_NUM != 502

#if !defined(lua_number2unsigned)       /* { */
/* the following definition assures proper modulo behavior */
#if defined(LUA_NUMBER_DOUBLE)
#include <math.h>
#define SUPUNSIGNED     ((lua_Number)(~(lua_Unsigned)0) + 1)
#define lua_number2unsigned(i,n)  \
        ((i)=(lua_Unsigned)((n) - floor((n)/SUPUNSIGNED)*SUPUNSIGNED))
#else
#define lua_number2unsigned(i,n)        ((i)=(lua_Unsigned)(n))
#endif
#endif                          /* } */

/* on several machines, coercion from unsigned to double is slow,
   so it may be worth to avoid */
#define lua_unsigned2number(u)  \
    (((u) <= (lua_Unsigned)INT_MAX) ? (lua_Number)(int)(u) : (lua_Number)(u))

/* ----- adapted from lua-5.2.0 lapi.c: ----- */

static void lua_pushunsigned (lua_State *L, lua_Unsigned u) {
  lua_Number n;
  n = lua_unsigned2number(u);
  lua_pushnumber(L, n);
}

/* ----- adapted from lua-5.2.0-work3 lbitlib.c getuintarg(): ----- */

static lua_Unsigned luaL_checkunsigned (lua_State *L, int arg) {
  lua_Unsigned r;
  lua_Number x = lua_tonumber(L, arg);
  if (x == 0) luaL_checktype(L, arg, LUA_TNUMBER);
  lua_number2unsigned(r, x);
  return r;
}

/* -----End of copied from bit32.c ----- */

#endif


#include "crc32.h"


static const char *crc32_reg_name      = "ehj.crc32";

static void create_meta(lua_State *L, const char *name, const luaL_Reg *lib) {
	luaL_newmetatable(L, name);
	lua_pushstring(L, "__index");
	lua_pushvalue(L, -2);               /* push metatable */
	lua_rawset(L, -3);                  /* metatable.__index = metatable */

	/* register metatable functions */
#if LUA_VERSION_NUM < 502
	luaL_register(L, NULL, lib);
#else
	luaL_setfuncs(L, lib, 0);
#endif

	/* remove metatable from stack */
	lua_pop(L, 1);
}

static int Lcrc_reset(lua_State *L) {
	uint32_t *_crc32 = (uint32_t*) luaL_checkudata(L, 1, crc32_reg_name);
	*_crc32 = 0;
	lua_settop(L, 1); //  return userdata itself
	return 1;
}
static int Lcrc_update(lua_State *L) {
	uint32_t *_crc32 = (uint32_t*) luaL_checkudata(L, 1, crc32_reg_name);
	const uint8_t *data;
	size_t data_len;
	
	luaL_checktype(L, 2, LUA_TSTRING);
	data = lua_tolstring (L, 2, &data_len);
	*_crc32 = crc32(*_crc32, data, data_len );
	lua_settop(L, 1); //  return userdata itself
	return 1;
}

static int Lcrc_tonumber(lua_State *L) {
	uint32_t *_crc32 = (uint32_t*) luaL_checkudata(L, 1, crc32_reg_name);
	lua_pushunsigned(L, *_crc32);
	return 1;
}

static int Lcrc_tostring(lua_State *L) {
	uint32_t *_crc32 = (uint32_t*) luaL_checkudata(L, 1, crc32_reg_name);
	uint32_t crc_out = *_crc32;
	uint8_t crcout_str[4];
	int i;
	
	for (i = 3; i>=0; i--) {
		crcout_str[i] = crc_out & 0xff;
		crc_out>>=8;
	}

	lua_pushlstring (L, crcout_str, 4);
	return 1;
}

static int Lcrc_tohex(lua_State *L) {
	uint32_t *_crc32 = (uint32_t*) luaL_checkudata(L, 1, crc32_reg_name);
	uint8_t crcout_str[9];
	sprintf(crcout_str, "%8.8x", *_crc32);
	lua_pushlstring (L, crcout_str, 8);
	return 1;
}

static const luaL_Reg crc32_meta_reg[] = {
	{"reset",      Lcrc_reset    },
	{"update",     Lcrc_update   },
	{"tonumber",   Lcrc_tonumber },
	{"tostring",   Lcrc_tostring },
	{"tohex",      Lcrc_tohex    },
	{NULL, NULL}
};


static int Lcrc32(lua_State *L) {
	uint32_t  crc_in, crc_out;
	const uint8_t *data;
	size_t data_len;
	
	luaL_checktype(L, 2, LUA_TSTRING);
	data = lua_tolstring (L, 2, &data_len);

	if (lua_type(L,1) == LUA_TSTRING) { 
		// input CRC as 4 byte string
		size_t crc_len;
		int i;
		const uint8_t *crcin_str = lua_tolstring(L,1, &crc_len);
		if (crc_len>4) crc_len = 4;
		crc_in = 0;
		for (i = 0; i < crc_len; i++) {
			crc_in<<=8;
			crc_in |=  crcin_str[i];
		}

		crc_out = crc32(crc_in, data, data_len );
			
		// return CRC as 4 byte string
		uint8_t crcout_str[4];

		for (i = 3; i>=0; i--) {
			crcout_str[i] = crc_out & 0xff;
			crc_out>>=8;
		}

		lua_pushlstring (L, crcout_str, 4);
	} else {
		crc_in = luaL_checkunsigned(L, 1);
		crc_out = crc32(crc_in, data, data_len );
		lua_pushunsigned(L, crc_out);
	}
	
	return 1;
}

static int Lnewcrc32(lua_State *L) {
	uint32_t *crc32 = (uint32_t *)lua_newuserdata(L, sizeof(uint32_t));
	if (luaL_newmetatable(L, crc32_reg_name)) {
		lua_pushstring(L, "__index");
		lua_pushvalue(L, -2);               /* push metatable */
		lua_rawset(L, -3);                  /* metatable.__index = metatable */

		/* register metatable functions */
#if LUA_VERSION_NUM < 502
		luaL_register(L, NULL, crc32_meta_reg);
#else
		luaL_setfuncs(L, crc32_meta_reg, 0);
#endif
	}
	lua_setmetatable(L, -2);
	*crc32 = 0;
	return 1;
}

static const luaL_Reg Rcrc32[] = {
	{"crc32",     Lcrc32     },
	{"newcrc32",  Lnewcrc32  },
	{NULL, NULL}
};

LUALIB_API int luaopen_crc32(lua_State *L) {
	create_meta(L,crc32_reg_name, crc32_meta_reg);
	
#if LUA_VERSION_NUM < 502
	lua_newtable (L);
	luaL_register(L, NULL, Rcrc32);
#else
	luaL_newlib(L, Rcrc32);
#endif

	lua_pushliteral(L, VERSION);
	lua_setfield(L, -2, "version");
	return 1;
}
